gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -743,6 +743,49 @@ protected GoogleCloudStorageItemInfo getInitialMetadata() {
     };
   }
 
+  public SeekableByteChannel open(
+          final GoogleCloudStorageItemInfo itemInfo, GoogleCloudStorageReadOptions readOptions)
+          throws IOException {
+    logger.atFiner().log("open(%s, %s)", itemInfo.getResourceId(), readOptions);
+    Preconditions.checkArgument(
+            itemInfo.getResourceId().isStorageObject(), "Expected full StorageObject id, got %s", itemInfo.getResourceId());
+
+    if (storageOptions.isGrpcEnabled()) {
+      return GoogleCloudStorageGrpcReadChannel
+              .open(storageStubProvider, storage, errorExtractor, itemInfo, readOptions);
+    }
+
+    // The underlying channel doesn't initially read data, which means that we won't see a
+    // FileNotFoundException until read is called. As a result, in order to find out if the object
+    // exists, we'll need to do an RPC (metadata or data). A metadata check should be a less
+    // expensive operation than a read data operation.
+//    GoogleCloudStorageItemInfo info;
+    if (readOptions.getFastFailOnNotFound()) {
```

Comment:
Can we move this check to the top of the method

Reply:
Is it called later because of grpc vs non-grpc channel?

Reply:
In this method, we have access to itemInfo, so we can failFast first